libs/cardano-protocol-tpraos/testlib/Test/Cardano/Protocol/TPraos/Create.hs

```diff
@@ -0,0 +1,294 @@
+  AllIssuerKeys c r ->
+  Word ->
+  -- | KES period of key registration
+  Word ->
+  BHBody c ->
+  BHeader c
+mkBHeader pKeys kesPeriod keyRegKesPeriod bhBody =
+  let sHot = kesSignKey $ snd $ NE.head $ aikHot pKeys
+      kpDiff = kesPeriod - keyRegKesPeriod
+      hotKey = case evolveKESUntil sHot (KESPeriod 0) (KESPeriod kpDiff) of
+        Nothing ->
+          error $
+            mconcat
+              [ "Could not evolve key to iteration. "
+              , "keyRegKesPeriod: " ++ show keyRegKesPeriod
+              , "kesPeriod: " ++ show kesPeriod
+              , "kpDiff: " ++ show kpDiff
+              ]
+        Just hKey -> hKey
+      sig = signedKES () kpDiff bhBody hotKey
+   in BHeader bhBody sig
+
+-- | Try to evolve KES key until specific KES period is reached, given the
+-- current KES period.
+evolveKESUntil ::
+  (KES.KESAlgorithm v, KES.ContextKES v ~ ()) =>
+  KES.SignKeyKES v ->
+  -- | Current KES period
+  KESPeriod ->
+  -- | Target KES period
+  KESPeriod ->
+  Maybe (KES.SignKeyKES v)
+evolveKESUntil sk1 (KESPeriod current) (KESPeriod target) = go sk1 current target
+  where
+    go !_ c t | t < c = Nothing
+    go !sk c t | c == t = Just sk
+    go !sk c t = case KES.updateKES () sk c of
+      Nothing -> Nothing
+      Just sk' -> go sk' (c + 1) t
+
+mkBlock ::
+  forall era r.
+  ( EraSegWits era
+  , VRF.Signable (VRF (EraCrypto era)) Seed
+  , KES.Signable (KES (EraCrypto era)) (BHBody (EraCrypto era))
+  ) =>
+  -- | Hash of previous block
+  HashHeader (EraCrypto era) ->
+  -- | All keys in the stake pool
+  AllIssuerKeys (EraCrypto era) r ->
+  -- | Transactions to record
+  [Tx era] ->
+  -- | Current slot
+  SlotNo ->
+  -- | Block number/chain length/chain "difficulty"
+  BlockNo ->
+  -- | EpochNo nonce
+  Nonce ->
+  -- | Period of KES (key evolving signature scheme)
+  Word ->
+  -- | KES period of key registration
+  Word ->
+  -- | Operational certificate
+  OCert (EraCrypto era) ->
+  Block (BHeader (EraCrypto era)) era
+mkBlock prev pKeys txns slotNo blockNo enonce kesPeriod keyRegKesPeriod oCert =
+  let protVer = ProtVer (eraProtVerHigh @era) 0
+      txseq = toTxSeq @era (StrictSeq.fromList txns)
+      bodySize = fromIntegral $ bBodySize protVer txseq
+      bodyHash = hashTxSeq @era txseq
+      bhBody = mkBHBody protVer prev pKeys slotNo blockNo enonce oCert bodySize bodyHash
+      bHeader = mkBHeader pKeys kesPeriod keyRegKesPeriod bhBody
+   in Block bHeader txseq
+
+-- | Create a block with a faked VRF result.
+mkBlockFakeVRF ::
+  forall era r.
+  ( EraSegWits era
+  , VRF.Signable (VRF (EraCrypto era)) (WithResult Seed)
+  , KES.Signable (KES (EraCrypto era)) (BHBody (EraCrypto era))
+  ) =>
+  -- | Hash of previous block
+  HashHeader (EraCrypto era) ->
+  -- | All keys in the stake pool
+  AllIssuerKeys (EraCrypto era) r ->
+  -- | Transactions to record
+  [Tx era] ->
+  -- | Current slot
+  SlotNo ->
+  -- | Block number\/chain length\/chain "difficulty"
```

Comment:
difficulty?

Reply:
I have no idea what it means. I didn't write it :smile:

Reply:
:laughing: , gotcha. I think "Block number" is i pretty good explanation myself :)